internal/repository/index.go

```diff
@@ -48,11 +48,10 @@ import (
 
 // Index holds lookup tables for id -> pack.
 type Index struct {
-	m          sync.Mutex
-	blob       map[restic.BlobHandle]indexEntry
-	duplicates map[restic.BlobHandle][]indexEntry
-	packs      restic.IDs
-	treePacks  restic.IDs
+	m         sync.Mutex
+	byType    map[restic.BlobType]*lowMemMap
```

Comment:
This little map may well be a performance killer on the benchmark. In #2812, I inline two lazily-initialized header structures in the Index.

Reply:
It didn't contribute too much to the performance. However, using `byType []*lowMemMap` is the way better option. This can be directly accessed with `byType[tpe]` where `tpe` is `restic.BlobType`.
I changed the implementation, so that you can use this as reference for modifying your PR (also made a comment there).